### PR TITLE
Automated test of infrastructure

### PR DESCRIPTION
**Pull request by fwilhe**

### CODE REVIEW

**Comment by fwilhe** [comment]
Still some `fixme` comments, but in general ready for review

**Comment by marcusholl** [comment]
Hi @fwilhe 

I tried to run the example ( `./runTests.sh`). I defined the environment variables and I get:

```
+ export PPIPER_INFRA_IT_CF_PASSWORD
+ export PPIPER_INFRA_IT_CF_USERNAME
+ chmod +x cx-server
+ ./cx-server start
invalid reference format
docker: invalid reference format.
```

is this a known issue?

**Comment by fwilhe** [comment]
Hi @marcusholl, this is not a known issue yet. Please share more details (in a private conversation if any sensitive information is in there) about your setup. Which OS are you using? Did you follow the steps described in the README?

**Comment by fwilhe** [comment]
Hello @SAP/devops-docker-images-team,

I know this is not perfect yet, but the remaining FIXMEs are not so quick to solve and we need the baseline test asap. I highly appreciate your reviews, with the goal to get this merged latest wednesday this week.

Thanks

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+language: minimal
+services:
+- docker
+script: cd infrastructure-tests && chmod +x runTests.sh && ./runTests.sh
+env:
+  global:
+  - PPIPER_INFRA_IT_TEST_PROJECT=https://github.com/SAP/cloud-s4-sdk-book
+  - PPIPER_INFRA_IT_CF_USERNAME=P2001099548
```
Comment (daniel-kurzynski) [inline .travis.yml]: Not sure if the P-User should be part of this repo, should be configured together with the password in travis

File: infrastructure-tests/Jenkinsfile [hunk 2]
```diff
@@ -0,0 +1,32 @@
+#!/usr/bin/env groovy
+import com.sap.piper.JenkinsController
+
+@Library(['s4sdk-pipeline-library', 'piper-library-os']) _
```
Comment (daniel-kurzynski) [inline infrastructure-tests/Jenkinsfile]: We should get rid of s4sdk-pipeline-library

File: infrastructure-tests/runTests.sh [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+set -x
+
+# Start a local registry, to which we push the images built in this test, and from which they will be consumed in the test
+docker run -d -p 5000:5000 --restart always --name registry registry:2 || true
+find ../cx-server-companion -type f -exec sed -i "" -e 's/ppiper/localhost:5000\/ppiper/g' {} \;
+
+# Configure our testing Jenkins instance via Configuration as Code (Create build job, register shared libraries, configure executors)
+mkdir -p ../cx-server-companion/life-cycle-scripts/jenkins-configuration
+cp testing-jenkins.yml ../cx-server-companion/life-cycle-scripts/jenkins-configuration
```
Comment (daniel-kurzynski) [inline infrastructure-tests/runTests.sh]: Might, took me a few seconds to understand that you are copying it there to configure the cx-server

File: infrastructure-tests/runTests.sh [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+docker build -t localhost:5000/ppiper/container-structure-test:latest ../container-structure-test
```
Comment (daniel-kurzynski) [inline infrastructure-tests/runTests.sh]: Why do you build it? We do not have tests for it...

File: infrastructure-tests/runTests.sh (hunk 3 above)
Comment (daniel-kurzynski) [reply infrastructure-tests/runTests.sh]: I would rather copy the cx-server here.

File: infrastructure-tests/runTests.sh [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+docker build -t localhost:5000/ppiper/jenkins-master:latest ../jenkins-master
+
+# The cx-server script will be updated to the version where this URI points to.
+#If you need to change it in a PR to work, change it here temporarily and move it afterwards back to upstream.
+docker build --build-arg cx_server_base_uri=https://raw.githubusercontent.com/sap/devops-docker-images/master/jenkins-master/cx-server/cx-server -t localhost:5000/ppiper/cx-server-companion:latest ../cx-server-companion
```
Comment (daniel-kurzynski) [inline infrastructure-tests/runTests.sh]: Do we still need this build arg. I thought we get rid of it...

File: infrastructure-tests/runTests.sh (hunk 4 above)
Comment (daniel-kurzynski) [reply infrastructure-tests/runTests.sh]: You also do not use it later on

File: cx-server-companion/life-cycle-scripts/cx-server [hunk 6]
```diff
@@ -27,6 +27,8 @@ fi
         --env DEVELOPER_MODE \
         --env host_os=unix \
         --env cx_server_path="${PWD}" \
+	--env PPIPER_INFRA_IT_CF_PASSWORD \
```
Comment (daniel-kurzynski) [inline cx-server-companion/life-cycle-scripts/cx-server]: would be in favor of a more generic approach, e.g. configure somewhere a list of environment variables

File: .travis.yml (hunk 1 above)
Comment (fwilhe) [reply .travis.yml]: This is the configuration in travis. The password is also in this file (encrypted). Not sure if travis provides another way to configure this.

File: infrastructure-tests/runTests.sh (hunk 3 above)
Comment (fwilhe) [reply infrastructure-tests/runTests.sh]: I don't see why this would be easier to understand.

File: infrastructure-tests/runTests.sh (hunk 4 above)
Comment (fwilhe) [reply infrastructure-tests/runTests.sh]: leftover, thanks.

File: infrastructure-tests/runTests.sh (hunk 5 above)
Comment (fwilhe) [reply infrastructure-tests/runTests.sh]: Well, it is all fine if your PR does not change cx-server, but if it does, you need a way to not overwrite it with the published file, ight

File: .travis.yml (hunk 1 above)
Comment (daniel-kurzynski) [reply .travis.yml]: I can also configure it in the UI of Travis ...

File: infrastructure-tests/runTests.sh (hunk 5 above)
Comment (daniel-kurzynski) [reply infrastructure-tests/runTests.sh]: The build.arg isn't even an build arg anymore:
https://github.com/SAP/devops-docker-images/blob/master/cx-server-companion/Dockerfile

I will always pick the files located in life-cycle-scripts/*

File: infrastructure-tests/runTests.sh (hunk 3 above)
Comment (daniel-kurzynski) [reply infrastructure-tests/runTests.sh]: For example it will also put the jenkins configuration into the companion image. Not sure if this can lead to side effects. 
I think it is cleaner to copy them into some temporary folder as the other folder is part of a docker image.

File: cx-server-companion/files/cx-server-companion.sh [hunk 7]
```diff
@@ -590,6 +591,10 @@ function start_jenkins_container()
         if [ -e /cx-server/mount/jenkins-configuration ]; then
             environment_variable_parameters+=(-e CASC_JENKINS_CONFIG=/var/cx-server/jenkins-configuration)
         fi
+
+        environment_variable_parameters+=(-e PPIPER_INFRA_IT_CF_PASSWORD)
+        environment_variable_parameters+=(-e PPIPER_INFRA_IT_CF_USERNAME)
```
Comment (radsoulbeard) [inline cx-server-companion/files/cx-server-companion.sh]: Is the cx-server-companion part of the infrastructure tests or also something productive?
I wonder why parameters from the tests are mentioned in the script.

File: cx-server-companion/life-cycle-scripts/cx-server (hunk 6 above)
Comment (daniel-kurzynski) [reply cx-server-companion/life-cycle-scripts/cx-server]: As @radsoulbeard said: As it is productive code and the content of this file is shown to the user I would like to see another approach which is nearer to productive uses cases (e.g. using config as code and configuring credentials productively)

File: cx-server-companion/files/cx-server-companion.sh (hunk 7 above)
Comment (fwilhe) [reply cx-server-companion/files/cx-server-companion.sh]: This is the productive life cycle mgmt script, but I don't see how we can get by without passing the environment for testing. Also, I don't see a big issue there. We need a way to instrument our environment for testing.

Comment (daniel-kurzynski) [reply cx-server-companion/files/cx-server-companion.sh]: Yes it is productive code. But here you can argue that it is not show to the user but hidden inside the docker container as a private API.

Comment (radsoulbeard) [reply cx-server-companion/files/cx-server-companion.sh]: Ok, thanks for clarification

File: cx-server-companion/files/cx-server-companion.sh [hunk 8]
```diff
@@ -590,6 +591,11 @@ function start_jenkins_container()
         if [ -e /cx-server/mount/jenkins-configuration ]; then
             environment_variable_parameters+=(-e CASC_JENKINS_CONFIG=/var/cx-server/jenkins-configuration)
         fi
+
+        for var in $(env | grep ^CX) do
+            environment_variable_parameters+=(-e $var)
```
Comment (radsoulbeard) [inline cx-server-companion/files/cx-server-companion.sh]: I get an syntax error in that line :-(

Comment (fwilhe) [reply cx-server-companion/files/cx-server-companion.sh]: Yes, me too, but I don't get it yet.. What is wrong?

Comment (rkamath3) [reply cx-server-companion/files/cx-server-companion.sh]:
```
for x in y
do
      # do something
done
```
or
```
for x in y; do
      # do something
done
```

File: cx-server-companion/files/cx-server-companion.sh [hunk 9]
```diff
@@ -590,6 +591,11 @@ function start_jenkins_container()
         if [ -e /cx-server/mount/jenkins-configuration ]; then
             environment_variable_parameters+=(-e CASC_JENKINS_CONFIG=/var/cx-server/jenkins-configuration)
         fi
+
+        for var in $(env | grep ^CX) do
```
Comment (rkamath3) [inline cx-server-companion/files/cx-server-companion.sh]: Why is this required?

File: infrastructure-tests/runTests.sh [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+if [ -z "$CX_INFRA_IT_CF_PASSWORD" ]; then
+    echo "Failure: Variable CX_INFRA_IT_CF_PASSWORD is unset"
+    exit 1
+fi
+
+set -x
+
+# Start a local registry, to which we push the images built in this test, and from which they will be consumed in the test
+docker run -d -p 5000:5000 --restart always --name registry registry:2 || true
+find ../cx-server-companion -type f -exec sed -i "" -e 's/ppiper/localhost:5000\/ppiper/g' {} \;
```
Comment (rkamath3) [inline infrastructure-tests/runTests.sh]: why do we need this? what is the desired behavior?

Comment (fwilhe) [reply infrastructure-tests/runTests.sh]: We need this to use the images we've built locally so they are consumed from our local registry, not from docker hub.

Comment (rkamath3) [reply infrastructure-tests/runTests.sh]: `find ../cx-server-companion -type f -exec sed -i -e 's/ppiper/localhost:5000\/ppiper/g' {} \;` ? empty string after `-i` tries to read the file and throws error `sed: can't read : No such file or directory`

Comment (fwilhe) [reply infrastructure-tests/runTests.sh]: Seems to work in travis, and that is all I can check

File: cx-server-companion/files/cx-server-companion.sh [hunk 11]
```diff
@@ -593,7 +593,7 @@ function start_jenkins_container()
         fi
 
         # Pass custom environment variables prefixed by 'CX', may be used to pass values into Configuration as Code
-        for var in $(env | grep ^CX)
+        for var in $(env | grep ^CX | cut -d'=' -f1)
```
Comment (rkamath3) [inline cx-server-companion/files/cx-server-companion.sh]: That's a good catch

File: infrastructure-tests/pipeline_config.yml [hunk 12]
```diff
@@ -0,0 +1,2 @@
+general:
+  productiveBranch: master
```
Comment (rkamath3) [inline infrastructure-tests/pipeline_config.yml]: newline
